amqp/src/main/scala/akka/stream/alpakka/amqp/impl/AmqpAsyncFlow.scala

```diff
@@ -0,0 +1,332 @@
+          log.debug("Pushing enqueued {} downstream.", result)
+          push(out, result)
+        }
+
+        if (isFinished) closeStage()
+        else if (!hasBeenPulled(in)) tryPull(in)
+      }
+    }
+  )
+
+  override protected def onTimer(timerKey: Any): Unit =
+    timerKey match {
+      case tag: DeliveryTag => {
+        log.debug("Received timeout for deliveryTag {}.", tag)
+        onRejection(tag, multiple = false)
+      }
+      case _ => ()
+    }
+
+  private def closeStage(): Unit =
+    upstreamException match {
+      case Some(throwable) =>
+        promise.failure(throwable)
+        failStage(throwable)
+      case None =>
+        promise.success(Done)
+        completeStage()
+    }
+
+  private def isFinished: Boolean = isClosed(in) && noAwaitingMessages && exitQueue.isEmpty
+}
+
+/**
+ * Internal API.
+ *
+ * AMQP flow that uses asynchronous confirmations but preserves the order in which messages were pulled.
+ * Internally messages awaiting confirmation are stored in a ordered buffer. Initially messages have `ready`
+ * flag set to `false`. On confirmation flag is changed to `true` and messages from the beginning of the queue
+ * are dequeued in sequence until first non-ready message - this means that if confirmed message is not a
+ * first element of the buffer then nothing will be dequeued. Flag `multiple` on a confirmation means that
+ * broker confirms all messages up to a given delivery tag, which means that so all messages up to (and including)
+ * this delivery tag can be safely dequeued.
+ */
+@InternalApi private[amqp] final class AmqpAsyncFlow[T](
+    sinkSettings: AmqpWriteSettings,
+    bufferSize: Int,
+    confirmationTimeout: FiniteDuration
+) extends GraphStageWithMaterializedValue[FlowShape[WriteMessage[T], WriteResult[T]], Future[Done]] {
+
+  val in: Inlet[WriteMessage[T]] = Inlet(Logging.simpleName(this) + ".in")
+  val out: Outlet[WriteResult[T]] = Outlet(Logging.simpleName(this) + ".out")
+
+  override def shape: FlowShape[WriteMessage[T], WriteResult[T]] = FlowShape.of(in, out)
+
+  override protected def initialAttributes: Attributes =
+    super.initialAttributes and Attributes.name(Logging.simpleName(this)) and ActorAttributes.IODispatcher
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[Done]) = {
+    val promise = Promise[Done]()
+    (new AmqpAsyncFlowStageLogic(sinkSettings, bufferSize, confirmationTimeout, promise, shape) {
+
+      private var buffer = TreeMap[DeliveryTag, AwaitingMessage[T]]()
+
+      override def enqueueMessage(tag: DeliveryTag, passThrough: T): Unit =
+        buffer += (tag -> AwaitingMessage(tag, passThrough))
+
+      override def dequeueAwaitingMessages(tag: DeliveryTag, multiple: Boolean): Iterable[AwaitingMessage[T]] = {
+        setReady(tag)
+
+        if (multiple) {
+          dequeueWhile((t, _) => t <= tag)
+        } else {
+          if (isAtHead(tag)) {
```

Comment:
could be just plain if-else-if-else, not nested.

Reply:
`setReady(tag)` should be called before the second `if`, it doesn't have to be called for `multiple` branch - I misplaced it during refactoring. So it can't be simplified as you suggested.